AL Income documentation

BryceStevenWilley opened the pull request:
Still in progress, but you can see the general shape of how I plan on writing the documentation pages in the ALExpenseList page. 

Polish still needed on the ALExpenseList page:
* doing a quick run through of the tutorial myself as a newer DA user

Additions needed for the rest of this PR in order of priority:
* ALItemizedJobList
* ALAssetList
* ALIncomeList

BryceStevenWilley commented:
@nonprofittechy and @plocket, I've added you as reviewers for the moment to make sure that the general approach of a tutorial like this would be useful and get any ideas y'all might have to improve said tutorial. I'll ping you both again when I've written all of the tutorials for ALIncome and change this from a draft to a ready PR then.

plocket commented on docs/alincome/expenses.md:
```diff
@@ -0,0 +1,271 @@
+slug: /alincome/expenses
+---
+
+The `ALExpenseList` is a class that lets you ask a user about their different expenses in a simple way.
```
I've noticed you've used the word 'simple' a couple times. You were trying to avoid that, right?

BryceStevenWilley replied:
Good catch. I was trying to avoid using it to imply that some coding process was simple, like "Simply change the docassemble configuration or something".  Making things simple (and straight-forward) for our users is an important goal of our libraries though, so it feels valid here? Open to suggestions, but can think of any improvements myself that don't make this sentence longer than it needs to be.

plocket replied:
"Simple" here seems a bit over confident, but I see what you're saying. I thought I caught a use of the word lower down being used to express that something was simple to do. I'll have to try and track that down.

```suggestion
The `ALExpenseList` is a class that helps you ask a user about their different expenses.
```

plocket commented on docs/alincome/expenses.md:
```diff
@@ -0,0 +1,271 @@
+Certain court forms ask for a person's expenses as a proof of hardship or burden.
+In designing this feature, we tried to reduce the number of questions a user needs
+to answer while still asking for specific enough expenses. We don't want to burden the user, but also
+doesn't want the user forget to list any significant expenses they would have
```
```suggestion
don't want the user forget to list any significant expenses they would have
```
To match the previous part of the sentence.

plocket commented on docs/alincome/expenses.md:
```diff
@@ -0,0 +1,271 @@
+Before you start, we'll assume that you:
+
+* have access to a [developer account on a docassemble server](https://suffolklitlab.org/legal-tech-class/docs/classes/assembly-line/2020-assembly-line-assignment-1#before-you-get-started)
+* have the [`ALToolbox` package installed on your server](https://suffolklitlab.org/docassemble-AssemblyLine-documentation/docs/installation#run-the-installation-script)
```
```suggestion
* have the [`ALToolbox` or `AssemblyLine` package installed on your server](https://suffolklitlab.org/docassemble-AssemblyLine-documentation/docs/installation#run-the-installation-script)
```
Many people will probably just be using AssemblyLine.

BryceStevenWilley replied:
Then they'll have `ALToolbox` already installed? I think I understand what you're pointing out; ppl might not immediately recognize the toolbox package, but they would know if they've installed the general assembly line. I might rephrase this differently to address that.

```suggestion
* have the [`AssemblyLine` package installed on your server](https://suffolklitlab.org/docassemble-AssemblyLine-documentation/docs/installation#run-the-installation-script), or have installed the [`ALToolbox` package](https://github.com/SuffolkLITLab/docassemble-ALToolbox) separately.
```

plocket commented on docs/alincome/expenses.md:
```diff
@@ -0,0 +1,271 @@
+
+  Sum of all of the expenses over a month: ${ currency(expenses.total(times_per_year=12)) }
+
+  Value of just one expense: ${ currency(expenses.total(source="rent")) }
```
I thought I remembered that multiple expenses could have the same source? If so, "Value of just one category of expense" and the description for "source" might need to be changed too. If not, never mind!

BryceStevenWilley replied:
Yes and no. Yes, `ALExpenseList` can have more than once expense from the same source. However the UX that we've landed on as the default (and is being used in this tutorial) won't have things from the same source, as it's showing a dict of checkboxes before hand, with each key for the checkbox as one source.

A user can select other and then add another expense from the same source, but TBH, that was more for resiliency and couldn't remember if there was a specific feature that someone would want to use with that.

Trying to balance being accurate and not confusing users with technical details they shouldn't have to know about to use the library is hard.  The description of source is accurate, and I don't think I would change it, even if we talk about allowing duplicates. I might change this line to be a bit more accurate though.

```suggestion
  Value of expenses with the source "rent": ${ currency(expenses.total(source="rent")) }
```

plocket commented on docs/alincome/jobs.md:
```diff
@@ -0,0 +1,191 @@
+
+### Writing the interview
+
+This interview will be simple; it will gather all of the information about someone's jobs, ask them to review it, and finally, will display the information.
```
Maybe this just needs a different word. Do you mean simple as in easy for a user to use, simple for the developer to use, or simple as in bare bones?

BryceStevenWilley replied:
Simple as in bare bones in this case. This is definitely a sub-par usage of the word, thanks.

plocket commented on docs/alincome/jobs.md:
```diff
@@ -0,0 +1,191 @@
+slug: /alincome/jobs
+---
+
+Asking people about their jobs is a common task when getting financial information. The `ALJob` class tries to be complete but simple to use and answer about, doing the
```
Simple for the developer and end user or just for the end user? e.g.
* "...but simple for you to use and simple for your users to answer..."?
* "...but simple for your users to complete..."?

BryceStevenWilley replied:
Simple for both, but this is another bad usage of the term. I'm being outed as adding the very thing I'm trying to remove, lol

plocket commented:
Is it worth adding a little bit to ALJobs/ALJobsList that mentions ALItemizedJob? In a way that describes the difference, so folks know have more context about when to use which one.

BryceStevenWilley commented:
I do already: https://github.com/SuffolkLITLab/docassemble-AssemblyLine-documentation/pull/310/files#diff-0d907de082071684a35b5c03fcc8804ddbc8fc896ebe76e0fb81489254f91428R80-R81. I plan on having a few more pointers to Itemized Jobs from Jobs, and a fuller comparison in `ALItemizedJob` though (not in `ALJob`, just to reduce up-front mental burden).

BryceStevenWilley commented:
@plocket , this is ready for review, I know you've been looking at it already, but feel free to do so again, I should have addressed your comments so far.

Not pinging nonprofittechy, since he's focused on other things this week.

plocket commented on docs/alincome/itemizedjobs.md:
```diff
@@ -5,16 +5,26 @@ title: |
-
 ## ALItemizedJob
 
+Sometimes, courts will ask for specific pieces of information about a person's pay, such as how many tips they received, or how much is deducted from their pay for health insurance. To keep track of all of those pieces of information for several jobs, you, the interview author,
```
Caroline has said that the main reason she wants this is because otherwise the user doesn't keep track of everything they need to take into account and they have to do a lot of the math themselves. You say a bit about that at the bottom of this section, but maybe it's worth thinking about which of those is the main purpose. For example, I think I understood that Caroline preferred it as a user experience over ALJob.

BryceStevenWilley replied:
Hm, if that's the case, there isn't a strong reason for ALJob at all? Should we still even document it / suggest it, or should we steer everyone towards ItemizedJob? 

I see what she means, but TBH I'm not sure if ALItemizedJob is there yet in terms of UX. I know that's the purpose of a lot of our other classes, like ALExpenseList, where we can ask the user about a lot of different expenses so they don't have to keep track of them. But ItemizedJob still feels cumbersome when doing that, and it doesn't quite remove the need for users to make calculations (a lot more discussion of that in https://github.com/SuffolkLITLab/docassemble-ALToolbox/issues/140).

BryceStevenWilley replied:
Now that I've thought about https://github.com/SuffolkLITLab/docassemble-ALToolbox/issues/140, I'm realizing that the core functionality of ALItemizedJob of handling additional income like tips, commissions, and overtime wages isn't flexible enough; all of those additional incomes are inconsistent, and asking about them in aggregate either require the user to do math (i.e. directly asking "what's your usual amount" or "what's the average"), or they just get over or under estimated.

I did a quick "look at this website I made" user test with my partner, and she pointed out some of these inconsistencies and unclear instructions. I'll make an issue for it separately, but I do think that the current language should probably stay for now, and maybe I'll add a warning with the above info as well.

Also will directly invite @CaroRob to weigh in if she has any thoughts.

plocket commented on docs/alincome/itemizedjobs.md:
```diff
@@ -25,7 +35,7 @@ Before you start, we'll assume that you
 * know [what the playground is](https://suffolklitlab.org/legal-tech-class/docs/classes/docacon-2020/hello-world#introduction-to-the-docassemble-playground) and [how to use it to develop a docassemble interview](https://suffolklitlab.org/legal-tech-class/docs/classes/docacon-2020/hello-world#hello-world)
 * know [what "blocks" are](https://suffolklitlab.org/legal-tech-class/docs/yaml#documents) in docassemble
-* have looked at the [the `ALJob` tutorial](/jobs) and decided you need to handle more complicated information
+* have looked at the [the `ALJob` tutorial](jobs.md) and decided you need to handle more complicated information
```
In aim of the above, maybe something like this would give fuller context:

```suggestion
* have looked at the [the `ALJob` tutorial](jobs.md) and decided you want to do the user's calculations for them or you need to handle more complicated information
```

plocket commented on docs/alincome/overview.md:
```diff
@@ -17,3 +17,23 @@ debts, and then allow users to review and add or change that information.
 
+If you want to jump right into the demo interview, you can install
+the [ALToolbox package](https://github.com/SuffolkLITLab/docassemble-ALTolbox)
+on your server, and run the [al_income_demo interview](https://github.com/SuffolkLITLab/docassemble-ALToolbox/blob/main/docassemble/ALToolbox/data/questions/al_income_demo.yml), by going to "https://yourserverurl.com/start/ALToolbox/al_income_demo". You can also visit out the live demo at [Suffolk's test server](https://apps-test.suffolklitlab.org/start/ALToolbox/al_income_demo).
```
I like that we have a demo up. Maybe list that first, since people (including me) often jump to do things instead of reading through to the end. Unless there's an advantage to them doing it instead.

plocket commented on docs/alincome/overview.md:
```diff
@@ -17,3 +17,23 @@ debts, and then allow users to review and add or change that information.
+
+:::caution Work in progress
+The following classes don't yet have tutorials that you can follow, but you can follow
+how they are used in the [al_income_demo interview](https://github.com/SuffolkLITLab/docassemble-ALToolbox/blobl/main/docassemble/ALToolbox/data/questions/al_income_demo.yml), and by checking out their source code below.
```
👍 I like this temporary alternative.

plocket commented:
Did you get to test all the code you showed as it's shown in the docs? Some of this is, I think, taken from parts of the demo interview, but have these pieces of code been independently tested? We could add them as hidden code to the demo interview and maybe just trigger the blocks when we need to make adjustments or for testing or something so we can know when things fall behind or if we've left something out. That's a heck of a lot of work, though. Maybe that could be a future issue.

plocket reviewed (commented):
I've not re-read the whole thing and I'm not sure what tomorrow will bring. For now, I'll just comment that this has been a good chunk of work and is at the very least a solid start. I think it's probably acceptable to merge. If I get to go through it again, I'll try to give it a more sure yay or nay.

plocket reviewed (approved):
Don't know what I'm on about! Take my other comments as you want. Approved unless I find something terribly wrong.

BryceStevenWilley commented:
> Did you get to test all the code you showed as it's shown in the docs?

Yes! I'll double test it before merging, in case I changed anything.

> We could add them as hidden code to the demo interview and maybe just trigger the blocks when we need to make adjustments or for testing or something so we can know when things fall behind or if we've left something out

I like the idea of automatically testing documentation, but if we do that I want a way to directly include the tested code into the documentation, which I don't know how to do right now. I'll make an issue for it.

BryceStevenWilley commented:
Waiting for https://github.com/SuffolkLITLab/docassemble-ALToolbox/pull/145 to be merged before merging here.
